lib/stitches/ripple_stitch.py

```diff
@@ -23,34 +25,100 @@ def ripple_stitch(stroke):
+    skip_end = _adjust_skip(stroke, num_lines, stroke.skip_end)
+
+    lines = _get_ripple_lines(stroke, helper_lines, is_linear, skip_start, skip_end)
+    stitches = _get_stitches(stroke, is_linear, lines)
 
     if stroke.reverse:
-        ripple_points.reverse()
+        stitches.reverse()
 
     if stroke.grid_size != 0:
-        ripple_points.extend(_do_grid(stroke, helper_lines))
-
-    stitches = running_stitch(ripple_points, stroke.running_stitch_length, stroke.running_stitch_tolerance)
+        stitches.extend(_do_grid(stroke, helper_lines, skip_start, skip_end))
 
     return _repeat_coords(stitches, stroke.repeats)
 
 
-def _do_ripple(stroke, helper_lines, is_linear):
-    points = []
+def _get_stitches(stroke, is_linear, lines):
+    if is_linear:
+        return _get_staggered_stitches(stroke, lines)
+    else:
+        points = [point for line in lines for point in line]
+        return running_stitch(points, stroke.running_stitch_length, stroke.running_stitch_tolerance)
+
+
+def _get_staggered_stitches(stroke, lines):
+    stitches = []
+    for i, line in enumerate(lines):
+        stitched_line = []
+        connector = []
+        if i != 0 and stroke.join_style == 0:
+            if i % 2 != 0:
+                last_point = lines[i-1][-1]
+                first_point = line[-1]
+            else:
+                last_point = lines[i-1][0]
+                first_point = line[0]
+            connector = running_stitch([InkstitchPoint(*last_point), InkstitchPoint(*first_point)],
+                                       stroke.running_stitch_length,
+                                       stroke.running_stitch_tolerance)
+        points = list(apply_stitches(LineString(line), stroke.running_stitch_length, stroke.staggers, 0.5, i, stroke.running_stitch_tolerance).coords)
+        stitched_line.extend([InkstitchPoint(*point) for point in points])
+        if i % 2 == 1:
+            # reverse every other row in linear ripple
+            stitched_line.reverse()
+        stitched_line = connector + stitched_line
+        stitches.extend(stitched_line)
+    return stitches
 
-    for point_num in range(stroke.get_skip_start(), len(helper_lines[0]) - stroke.get_skip_end()):
+
+def _adjust_skip(stroke, num_lines, skip):
+    if stroke.skip_start + stroke.skip_end >= num_lines:
+        return 0
+    return skip
+
+
+def _get_ripple_lines(stroke, helper_lines, is_linear, skip_start, skip_end):
+    lines = []
+    for point_num in range(skip_start, len(helper_lines[0]) - skip_end):
         row = []
         for line_num in range(len(helper_lines)):
             row.append(helper_lines[line_num][point_num])
+        lines.append(row)
+    return lines
 
-        if is_linear and point_num % 2 == 1:
-            # reverse every other row in linear ripple
-            row.reverse()
-
-        points.extend(row)
 
-    return points
+def get_line_count(stroke, pairs_or_line):
```

Comment:
This function makes my brain hurt too... maybe I shouldn't code late at night on vacation? ;)

I wonder if we should have a bunch of different get_line_count functions.  As it is now, it's a function that takes a bunch of different possible types and does a bunch of different things based on the types...